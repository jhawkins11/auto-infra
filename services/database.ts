import { InitAnswers } from '../types/prompts.js';
import {
  ensureDirectoryExists,
  writeConfigFile,
  writeSecretsFile,
} from './fileWriter.js';

const instanceSizeToClass: Record<
  NonNullable<InitAnswers['dbInstanceSize']>,
  string
> = {
  micro: 'db.t4g.micro',
  small: 'db.t4g.small',
  medium: 'db.t4g.medium',
  large: 'db.t4g.large',
  xlarge: 'db.t4g.xlarge',
};

export const generateDatabaseConfig = async (
  answers: InitAnswers,
): Promise<void> => {
  const configDirectory = `./${answers.projectName}`;
  const databaseConfigPath = `${configDirectory}/db.tfvars`;
  const secretsConfigPath = `${configDirectory}/secrets.db.tfvars`;

  const engine = answers.dbEngine ?? 'postgresql';
  const sizeKey = answers.dbInstanceSize ?? 'small';

  const nonSensitive = {
    project_name: answers.projectName,
    aws_region: answers.awsRegion,
    db_engine: engine,
    db_instance_class: instanceSizeToClass[sizeKey],
  } as const;

  const sensitive = {
    db_username: answers.dbUsername ?? 'admin',
    db_password: answers.dbPassword ?? '',
  } as const;

  await ensureDirectoryExists(databaseConfigPath);
  await writeConfigFile(
    databaseConfigPath,
    nonSensitive as unknown as Record<string, string | number | boolean>,
  );
  await writeSecretsFile(
    secretsConfigPath,
    sensitive as unknown as Record<string, string | number | boolean>,
  );
};
