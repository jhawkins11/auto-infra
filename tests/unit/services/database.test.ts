jest.mock('node:fs/promises', () => ({
  writeFile: jest.fn(),
  mkdir: jest.fn(),
}));

import * as fs from 'node:fs/promises';

import { generateDatabaseConfig } from '../../../services/database.js';
import { mustExist } from '../../utils/assertions.js';
import { createTestAnswers } from '../../utils/createTestAnswers.js';

describe('generateDatabaseConfig', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should write db.tfvars and secrets.db.tfvars files with correct content', async () => {
    const answers = createTestAnswers({
      projectName: 'db-project',
      awsRegion: 'us-east-2',
      dbEngine: 'postgresql',
      dbInstanceSize: 'small',
      dbUsername: 'dbadmin',
      dbPassword: 'supersecret',
    });

    await generateDatabaseConfig(answers);

    const mkdirMock = jest.mocked(fs.mkdir);
    const writeFileMock = jest.mocked(fs.writeFile);

    expect(mkdirMock).toHaveBeenCalledTimes(1);
    expect(mkdirMock).toHaveBeenCalledWith(`./${answers.projectName}`, {
      recursive: true,
    });

    const writeCalls = writeFileMock.mock.calls as Array<[string, string]>;
    const nonSecretCall = mustExist(
      writeCalls.find((c) => String(c[0]).endsWith('/db.tfvars')),
    );
    const secretCall = mustExist(
      writeCalls.find((c) => String(c[0]).endsWith('/secrets.db.tfvars')),
    );

    const [, nonSecretContent] = nonSecretCall as [string, string];
    const [, secretContent] = secretCall as [string, string];

    expect(nonSecretContent).toMatchSnapshot('db-non-sensitive');
    expect(secretContent).toMatchSnapshot('db-sensitive');
  });

  it('should apply defaults when engine/size are missing', async () => {
    const answers = createTestAnswers({ projectName: 'db-defaults' });
    // remove optional properties so defaults apply
    delete (answers as Partial<typeof answers>).dbEngine;
    delete (answers as Partial<typeof answers>).dbInstanceSize;

    await generateDatabaseConfig(answers);

    const writeFileMock = jest.mocked(fs.writeFile);
    const writeCalls = writeFileMock.mock.calls as Array<[string, string]>;
    const nonSecretCall = mustExist(
      writeCalls.find((c) => String(c[0]).endsWith('/db.tfvars')),
    );
    const [, nonSecretContent] = nonSecretCall as [string, string];

    expect(String(nonSecretContent)).toContain('db_engine = "postgresql"');
    expect(String(nonSecretContent)).toContain(
      'db_instance_class = "db.t4g.small"',
    );
  });
});
