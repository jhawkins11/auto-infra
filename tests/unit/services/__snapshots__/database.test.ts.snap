// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`generateDatabaseConfig should write db.tfvars and secrets.db.tfvars files with correct content: db-non-sensitive 1`] = `
"project_name = "db-project"
aws_region = "us-east-2"
db_engine = "postgresql"
db_instance_class = "db.t4g.small""
`;

exports[`generateDatabaseConfig should write db.tfvars and secrets.db.tfvars files with correct content: db-sensitive 1`] = `
"db_username = "dbadmin"
db_password = "supersecret""
`;
